parastell/magnet_coils.py

```diff
@@ -401,164 +324,105 @@ class MagnetCoil(object):
-
-        # Extract initial path point
-        pos = cubit.vertex(path_origin).coordinates()
-
-        # Project position vector onto cross-section
-        pos_proj = normalize(pos - np.inner(pos, norm) * norm)
-
-        # Compute angle by which to rotate cross-section such that it faces the
-        # origin
-        rot_ang_orig = np.arccos(np.inner(pos_proj, t_vec_rot))
-
-        # Re-orient rotated cross-section such that thickness vector faces
-        # origin
-        cubit.cmd(
-            f"rotate Surface {surf_id} angle {np.rad2deg(rot_ang_orig)} about "
-            "origin 0 0 0 direction " + " ".join(str(i) for i in norm)
+    @property
+    def coords(self):
+        return self._coords
+
+    @coords.setter
+    def coords(self, data):
+        # Compute tangents
+        tangents = np.subtract(
+            np.append(data[1:], [data[1]], axis=0),
+            np.append([data[-2]], data[0:-1], axis=0),
         )
+        tangents = tangents / np.linalg.norm(tangents, axis=1)[:, np.newaxis]
+
+        # Sample filament coordinates and tangents by modifier
+        self._coords = data[0 : -1 : self.sample_mod]
+        self._coords = np.append(self._coords, [self._coords[0]], axis=0)
+        self.tangents = tangents[0 : -1 : self.sample_mod]
+        self.tangents = np.append(self.tangents, [self.tangents[0]], axis=0)
 
     def create_magnet(self):
-        """Creates magnet coil volumes in cubit.
+        """Creates a single magnet coil CAD solid in CadQuery.
 
         Returns:
-            volume_id (int): magnet volume ids in cubit
+            coil (object): cq.Solid object representing a single magnet coil.
         """
-        # Cross-section inititally populated with thickness vector
-        # oriented along x axis
-        t_vec = np.array([1, 0, 0])
-
-        # Create cross-section for sweep
-        cubit.cmd(f"create surface " + self.shape_str + " zplane")
+        tangent_vectors = [
+            cq.Vector(tuple(tangent)) for tangent in self.tangents
+        ]
 
-        # Store cross-section index
-        cs_id = cubit.get_last_id("surface")
-        # Cross-section initially populated with normal oriented along z
-        # axis
-        cs_axis = np.array([0, 0, 1])
+        # Define coil filament path normals such that they face the filament
+        # center of mass
+        # Compute "outward" direction as difference between filament positions
+        # and filament center of mass
+        outward_dirs = self._coords - self.center_of_mass
+        outward_dirs = (
+            outward_dirs / np.linalg.norm(outward_dirs, axis=1)[:, np.newaxis]
+        )
 
-        # Initialize path list
-        path = []
+        # Project outward directions onto desired coil cross-section (CS) plane
+        # at each filament position to define filament path normals
+        parallel_parts = np.diagonal(
+            np.matmul(outward_dirs, self.tangents.transpose())
+        )
```

Comment:
See my comment elsewhere... would this get you a vector if you swapped the transpose?
```suggestion
        parallel_parts = p.matmul(outward_dirs.transpose(), self.tangents)
```
They are both 3 x N matrices, right? so there must be some way to arrange them such that a matrix multiplication results in a 1 x N vector without needing to take the diagonal...

Reply:
Perhaps I'm not seeing the same thing as you but I don't believe there's a way for a matrix product to be smaller in a dimension than the smallest dimension of the original matrices. I.e., for two Nx3 matrices, we can produce a 3x3 matrix (first matrix transposed) or a NxN matrix (second matrix transposed). That said, there may be some NumPy operation capable of producing what we want but I can't seem to find one...

Reply:
I really am getting old.... and would probably fail my qual...  I'll stop questioning these things when they work...